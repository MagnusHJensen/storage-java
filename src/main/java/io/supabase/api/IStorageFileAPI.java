package io.supabase.api;

import io.supabase.data.file.*;
import io.supabase.utils.MessageResponse;

import java.io.File;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public interface IStorageFileAPI {

    CompletableFuture<FilePathResponse> upload(String path, File file);
    CompletableFuture<FilePathResponse> update(String path, File file);
    CompletableFuture<MessageResponse> move(String fromPath, String toPath);
    CompletableFuture<FilePathResponse> copy(String fromPath, String toPath);
    CompletableFuture<FileSignedUrlResponse> getSignedUrl(String path, int expiresIn, FileDownloadOption options);
    CompletableFuture<List<FileSignedUrlResponse>> getSignedUrls(List<String> paths, int expiresIn, FileDownloadOption options);
    CompletableFuture<FileDownload> download(String path);
    FilePublicUrlResponse getPublicUrl(String path, FileDownloadOption options);

    CompletableFuture<List<io.supabase.data.file.File>> list(FileSearchOptions options);
    CompletableFuture<List<io.supabase.data.file.File>> list(String path, FileSearchOptions options);
    CompletableFuture<List<io.supabase.data.file.File>> delete(List<String> paths);
}
