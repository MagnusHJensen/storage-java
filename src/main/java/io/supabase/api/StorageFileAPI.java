package io.supabase.api;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import io.supabase.data.file.*;
import io.supabase.utils.HttpMethod;
import io.supabase.utils.MessageResponse;
import io.supabase.utils.RestUtils;
import io.supabase.utils.Utilities;
import okhttp3.Headers;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

public class StorageFileAPI implements IStorageFileAPI {
    private final String url;
    private final Map<String, String> headers;
    private final String bucketId;

    public StorageFileAPI(String url, Map<String, String> headers, String bucketId) {
        this.url = url;
        this.headers = headers;
        this.bucketId = bucketId;
    }

    private CompletableFuture<FilePathResponse> uploadOrUpdate(HttpMethod method, String path, File file) {
        InputStream fileStream;
        RequestBody requestBody;
        try {
            fileStream = Files.newInputStream(file.toPath());
            requestBody = Utilities.createRequestBody(MediaType.parse(Files.probeContentType(Paths.get(file.getPath()))), fileStream);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        String finalPath = getFinalPath(path);
        Request request = new Request.Builder().url(this.url + "object/" + finalPath).headers(Headers.of(this.headers)).method(method.getValue(), requestBody).build();
        return RestUtils.getCompletableFuture(request, new TypeToken<FilePathResponse>() {
        }.getType());
    }

    /**
     * <p>POST /object/{bucketName}/{wildcard}</p>
     * @param path The path to the file within the bucket of where it should be uploaded.
     * @param file The file that needs to be uploaded.
     * @return a {@link FilePathResponse}
     */
    @Override
    public CompletableFuture<FilePathResponse> upload(String path, File file) {
        return uploadOrUpdate(HttpMethod.POST, path, file);
    }

    /**
     * <p>PUT /object/{bucketName}/{wildcard}</p>
     * @param path The path to the file within the bucket that should get updated.
     * @param file The new file that should be placed.
     * @return a {@link FilePathResponse}
     */
    @Override
    public CompletableFuture<FilePathResponse> update(String path, File file) {
        return uploadOrUpdate(HttpMethod.PUT, path, file);
    }

    /**
     * <p>POST /object/move/</p>
     * @param fromPath The path to the object that needs to be moved.
     * @param toPath The new path where the object should be moved to.
     * @return a {@link MessageResponse}
     */
    @Override
    public CompletableFuture<MessageResponse> move(String fromPath, String toPath) {
        JsonObject body = new JsonObject();
        body.addProperty("bucketId", this.bucketId);
        body.addProperty("sourceKey", fromPath);
        body.addProperty("destinationKey", toPath);
        return RestUtils.post(new TypeToken<MessageResponse>() {
        }, headers, this.url + "object/move", body);
    }

    /**
     * <p>POST /object/copy/</p>
     * @param fromPath The path to the object that needs to be copied.
     * @param toPath The new path where the object should be copied to.
     * @return a {@link FilePathResponse}
     */
    @Override
    public CompletableFuture<FilePathResponse> copy(String fromPath, String toPath) {
        JsonObject body = new JsonObject();
        body.addProperty("bucketId", this.bucketId);
        body.addProperty("sourceKey", fromPath);
        body.addProperty("destinationKey", toPath);
        return RestUtils.post(new TypeToken<FilePathResponse>() {
        }, headers, this.url + "object/copy", body);
    }

    /**
     * <p>POST /object/sign/{bucketName}</p>
     * <p>This method just wraps {@link #getSignedUrls(List, int, FileDownloadOption, FileTransformOptions)}</p>
     *
     * @param path      The singular file path that should be signed.
     * @param expiresIn how many seconds until the signed url expires.
     * @param downloadOptions   any additional download options.
     * @param transformOptions The transform options if any
     * @return a {@link FileSignedUrlResponse}
     */
    @Override
    public CompletableFuture<FileSignedUrlResponse> getSignedUrl(String path, int expiresIn, FileDownloadOption downloadOptions, FileTransformOptions transformOptions) {
        CompletableFuture<List<FileSignedUrlResponse>> signedUrlFuture = getSignedUrls(List.of(path), expiresIn, downloadOptions, transformOptions);
        return signedUrlFuture.thenApply((responses) -> responses.get(0));
    }

    /**
     * <p>POST /object/sign/{bucketName}</p>
     *
     * @param paths     a list of file paths that should be signed.
     * @param expiresIn how many seconds until the signed urls expires.
     * @param downloadOptions   any additional download options.
     * @param transformOptions The transform options if any
     * @return a list of {@link FileSignedUrlResponse}
     */
    @Override
    public CompletableFuture<List<FileSignedUrlResponse>> getSignedUrls(List<String> paths, int expiresIn, FileDownloadOption downloadOptions, FileTransformOptions transformOptions) {
        JsonObject body = new JsonObject();
        JsonArray bodyPaths = new JsonArray();
        for (String path : paths) {
            bodyPaths.add(path);
        }
        body.addProperty("expiresIn", expiresIn);
        body.add("paths", bodyPaths);

        Map<String, String> downloadParamsMap = downloadOptions != null ? downloadOptions.convertToMap() : new HashMap<>();
        Map<String, String> transformParamsMap = transformOptions != null ? transformOptions.convertToMap() : new HashMap<>();
        Map<String, String> paramsMap = Utilities.combineMaps(downloadParamsMap, transformParamsMap);
        CompletableFuture<List<FileSignedUrlResponse>> signedUrlFuture = RestUtils.post(new TypeToken<List<FileSignedUrlResponse>>() {
        }, headers, this.url + "object/sign/" + bucketId, body);

        return signedUrlFuture.thenApply(fileSignedUrlResponse -> fileSignedUrlResponse.stream().map((fileSignedUrlResponse1 -> new FileSignedUrlResponse(this.url + fileSignedUrlResponse1.getSignedUrl() + Utilities.convertMapToQueryParams(paramsMap)))).collect(Collectors.toList()));
    }

    /**
     * <p>Downloads a file from a private bucket. To download something from a public bucket, make a request to the url from {@link #getPublicUrl(String, FileDownloadOption, FileTransformOptions)}</p>
     * <p>GET /object/authenticated/{bucketName}/{wildcard}</p>
     *
     * @param path The path of the file to download
     * @param transformOptions The transform options if any
     * @return a {@link FileDownload}
     */
    @Override
    public CompletableFuture<FileDownload> download(String path, FileTransformOptions transformOptions) {
        Request request = new Request.Builder().url(url + "object/authenticated/" + getFinalPath(path) + Utilities.convertMapToQueryParams(transformOptions.convertToMap())).get().headers(Headers.of(headers)).build();

        return RestUtils.getFile(request);
    }

    /**
     * Creates the url for an object in a public bucket
     *
     * @param path    The path of the file the link should point to.
     * @param downloadOptions The download options if any.
     * @param transformOptions The transform options if any
     * @return a {@link FilePublicUrlResponse}
     */
    @Override
    public FilePublicUrlResponse getPublicUrl(String path, FileDownloadOption downloadOptions, FileTransformOptions transformOptions) {
        boolean wantsTransformation = transformOptions != null;
        String urlSuffix = wantsTransformation ? "render/image" : "object";

        Map<String, String> transformParamsMap = transformOptions != null ? transformOptions.convertToMap() : new HashMap<>();
        Map<String, String> downloadParamsMap = downloadOptions != null ? downloadOptions.convertToMap() : new HashMap<>();
        Map<String, String> combined = Utilities.combineMaps(downloadParamsMap, transformParamsMap);

        return new FilePublicUrlResponse(this.url + urlSuffix + "/public/" + getFinalPath(path) + Utilities.convertMapToQueryParams(combined));
    }

    /**
     * <p>POST /object/list/{bucketName}</p>
     * @param options Options for the file search
     * @return a list of {@link io.supabase.data.file.File}
     */
    @Override
    public CompletableFuture<List<io.supabase.data.file.File>> list(FileSearchOptions options) {
        return list("", options);
    }

    /**
     * <p>POST /object/list/{bucketName}</p>
     * @param path The prefix to list objects by
     * @param options Options for the file search
     * @return a list of {@link io.supabase.data.file.File}
     */
    @Override
    public CompletableFuture<List<io.supabase.data.file.File>> list(String path, FileSearchOptions options) {
        JsonObject body = new Gson().toJsonTree(options).getAsJsonObject();
        body.addProperty("prefix", path);
        body.addProperty("search", "");
        return RestUtils.post(new TypeToken<List<io.supabase.data.file.File>>() {
        }, headers, this.url + "object/list/" + this.bucketId, body);
    }

    /**
     * <p>DELETE /object/{bucketName}</p>
     * @param paths a list of paths to files that should be deleted. BE WARY! PROVIDING NULL OR AN EMPTY LIST WILL DELETE THE ENTIRE BUCKET!
     * @return a list of the deleted files
     */
    @Override
    public CompletableFuture<List<io.supabase.data.file.File>> delete(List<String> paths) {
        JsonObject body = new JsonObject();
        JsonArray prefixes = new JsonArray();
        for (String path : paths) {
            prefixes.add(path);
        }
        body.add("prefixes", prefixes);
        return RestUtils.delete(new TypeToken<List<io.supabase.data.file.File>>() {
        }, headers, this.url + "object/" + bucketId, body);
    }

    private String getFinalPath(String path) {
        return bucketId + "/" + path;
    }
}
