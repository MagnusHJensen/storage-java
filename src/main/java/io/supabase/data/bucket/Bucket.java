package io.supabase.data.bucket;


import com.google.gson.annotations.JsonAdapter;
import com.google.gson.annotations.SerializedName;
import io.supabase.utils.FileSize;

import java.util.List;

public record Bucket(String id, String name, String owner,
                     @SerializedName("public") boolean isBucketPublic,
                     @SerializedName("file_size_limit") @JsonAdapter(FileSize.class) FileSize fileSizeLimit,
                     @SerializedName("allowed_mime_types") List<String> allowedMimeTypes,
                     @SerializedName("created_at") String createdAt,
                     @SerializedName("updated_at") String updatedAt) {

    @Override
    public String toString() {
        return "Bucket{" +
            "id='" + id + '\'' +
            ", name='" + name + '\'' +
            ", owner='" + owner + '\'' +
            ", isBucketPublic=" + isBucketPublic +
            ", createdAt='" + createdAt + '\'' +
            ", updatedAt='" + updatedAt + '\'' +
            '}';
    }
}
