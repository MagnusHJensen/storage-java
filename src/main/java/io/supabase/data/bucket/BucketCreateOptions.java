package io.supabase.data.bucket;

public class BucketCreateOptions extends BucketOptions {

    public BucketCreateOptions(boolean isPublic) {
        super(isPublic);
    }
}
