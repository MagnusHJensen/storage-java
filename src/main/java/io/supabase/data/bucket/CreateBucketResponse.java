package io.supabase.data.bucket;

public class CreateBucketResponse {
    public String name;

    public CreateBucketResponse(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return "CreateBucketResponse{" +
            "name='" + name + '\'' +
            '}';
    }
}
