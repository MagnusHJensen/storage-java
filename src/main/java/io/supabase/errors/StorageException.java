package io.supabase.errors;

public class StorageException extends Exception {
    private final String statusCode;
    private final String error;

    /**
     * Constructs a new error with the specified detail message.  The
     * cause is not initialized, and may subsequently be initialized by
     * a call to {@link #initCause}.
     *
     * @param message the detail message. The detail message is saved for
     *                later retrieval by the {@link #getMessage()} method.
     */
    public StorageException(String message, String statusCode, String error) {
        super(message);
        this.statusCode = statusCode;
        this.error = error;
    }

    public String getStatusCode() {
        return statusCode;
    }

    public String getError() {
        return error;
    }
}
