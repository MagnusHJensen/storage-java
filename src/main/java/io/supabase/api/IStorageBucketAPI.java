package io.supabase.api;

import io.supabase.data.bucket.Bucket;
import io.supabase.data.bucket.BucketCreateOptions;
import io.supabase.data.bucket.BucketUpdateOptions;
import io.supabase.data.bucket.CreateBucketResponse;
import io.supabase.utils.MessageResponse;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public interface IStorageBucketAPI {

    /**
     * <p>POST /bucket/</p>
     *
     * @param bucketId The name/id of the bucket you want to create.
     * @return a {@link CreateBucketResponse} object.
     */
    CompletableFuture<CreateBucketResponse> createBucket(String bucketId);

    /**
     * <p>POST /bucket/</p>
     *
     * @param bucketId The name/id of the bucket you want to create.
     * @param options  The options you want to pass for the bucket creation.
     * @return a {@link CreateBucketResponse} object.
     */
    CompletableFuture<CreateBucketResponse> createBucket(String bucketId, BucketCreateOptions options);

    /**
     * <p>GET /bucket/</p>
     * Lists all buckets in the project.
     *
     * @return A list of all buckets.
     */
    CompletableFuture<List<Bucket>> listBuckets();

    /**
     * <p>POST /bucket/{bucketId}/empty</p>
     *
     * @param bucketId The bucket you want to empty.
     * @return a status message
     */
    CompletableFuture<MessageResponse> emptyBucket(String bucketId);

    /**
     * <p>GET /bucket/{bucketId}/</p>
     *
     * @param bucketId The bucket of which you want to retrieve.
     * @return the asked for bucket
     */
    CompletableFuture<Bucket> getBucket(String bucketId);

    /**
     * <p>PUT /bucket/{bucketId}/</p>
     *
     * @param bucketId The bucket you want to update
     * @param options  The options to update the bucket to.
     * @return a status message
     */
    CompletableFuture<MessageResponse> updateBucket(String bucketId, BucketUpdateOptions options);

    /**
     * <p>DELETE /bucket/{bucketId}</p>
     *
     * @param bucketId The bucket of which you want to delete.
     * @return a status message
     */
    CompletableFuture<MessageResponse> deleteBucket(String bucketId);

}
