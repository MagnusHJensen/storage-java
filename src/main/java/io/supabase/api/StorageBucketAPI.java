package io.supabase.api;

import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import io.supabase.data.bucket.Bucket;
import io.supabase.data.bucket.BucketCreateOptions;
import io.supabase.data.bucket.BucketUpdateOptions;
import io.supabase.data.bucket.CreateBucketResponse;
import io.supabase.utils.MessageResponse;
import io.supabase.utils.RestUtils;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class StorageBucketAPI implements IStorageBucketAPI {

    /**
     * The API Key which is used to make requests on behalf of.
     */
    private final String apiKey;

    /**
     * <p>The url for your project.</p>
     * Example: {@code https://<PROJECT_ID>.supabase.co/storage/v1/}
     */
    private final String url;

    private final Map<String, String> headers;

    public StorageBucketAPI(String apiKey, String url) {
        this.apiKey = apiKey;
        this.url = url;
        this.headers = new HashMap<>();
        this.headers.put("Authorization", "Bearer " + apiKey);
    }


    @Override
    public CompletableFuture<CreateBucketResponse> createBucket(String bucketId) {
        return createBucket(bucketId, new BucketCreateOptions(false));
    }

    @Override
    public CompletableFuture<CreateBucketResponse> createBucket(String bucketId, BucketCreateOptions options) {
        JsonObject body = new JsonObject();
        body.addProperty("name", bucketId);
        body.addProperty("id", bucketId);
        body.addProperty("public", options.isPublic());
        return RestUtils.post(new TypeToken<CreateBucketResponse>(){}, headers, url + "bucket", body);
    }

    /**
     * Lists all buckets in the project.
     *
     * @return A list of all buckets.
     */
    @Override
    public CompletableFuture<List<Bucket>> listBuckets() {
        return RestUtils.get(new TypeToken<List<Bucket>>() {
        }, headers, url + "bucket");
    }

    @Override
    public CompletableFuture<MessageResponse> emptyBucket(String bucketId) {
        String urlPath = String.format("bucket/%s/empty", bucketId);
        return RestUtils.post(new TypeToken<MessageResponse>(){}, headers, url + urlPath, null);
    }

    @Override
    public CompletableFuture<Bucket> getBucket(String bucketId) {
        String urlPath = String.format("bucket/%s", bucketId);
        return RestUtils.get(new TypeToken<Bucket>(){}, headers, url + urlPath);
    }

    @Override
    public CompletableFuture<MessageResponse> updateBucket(String bucketId, BucketUpdateOptions options) {
        JsonObject body = new JsonObject();
        body.addProperty("public", options.isPublic());
        String path = String.format("bucket/%s", bucketId);
        return RestUtils.put(new TypeToken<MessageResponse>(){}, headers, url + path, body);
    }

    @Override
    public CompletableFuture<MessageResponse> deleteBucket(String bucketId) {
        String path = String.format("bucket/%s", bucketId);
        return RestUtils.delete(new TypeToken<MessageResponse>(){}, headers, url + path, null);
    }
}
