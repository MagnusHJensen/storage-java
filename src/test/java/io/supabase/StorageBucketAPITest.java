package io.supabase;


import io.supabase.data.bucket.Bucket;
import io.supabase.data.bucket.BucketCreateOptions;
import io.supabase.data.bucket.CreateBucketResponse;
import io.supabase.errors.StorageException;
import io.supabase.utils.MessageResponse;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.concurrent.ExecutionException;

import static org.junit.jupiter.api.Assertions.*;

public class StorageBucketAPITest {
    private static StorageClient client;

    @BeforeAll
    public static void initialize() throws ExecutionException, InterruptedException {
        client = new StorageClient("", "");
        client.createBucket("testbucket").get();
        // add files to bucket to make deleteBucket tests work.
    }

    @Test
    public void createBucket() throws ExecutionException, InterruptedException {
        CreateBucketResponse message = client.createBucket("firstbucket").get();
        client.createBucket("secondbucket", new BucketCreateOptions(true)).get();

        Bucket actual = client.getBucket("secondbucket").get();
        assertEquals(message.getName(), "firstbucket");
        assertTrue(actual.isBucketPublic());

        // cleanup
        client.deleteBucket("firstbucket").get();
        client.deleteBucket("secondbucket").get();
    }

    @Test
    public void listBuckets() throws ExecutionException, InterruptedException {
        List<Bucket> future = client.listBuckets().get();

        assertEquals(1, future.size());
    }

    @Test
    public void getBucket() throws ExecutionException, InterruptedException {
        Bucket future = client.getBucket("testbucket").get();

        assertNotNull(future);
    }

    @Test
    public void getBucketThrowsIfNotFound() {
        assertThrows(StorageException.class, () -> {
            try {
                client.getBucket("thisbucketcertainlydoesnotexist").get();
            } catch (ExecutionException e) {
                throw e.getCause();
            }
        }, "Not found");
    }

    @Test
    public void emptyBucket() throws ExecutionException, InterruptedException {
        MessageResponse message = client.emptyBucket("testbucket").get();

        assertEquals(message.getMessage(), "Successfully emptied");
    }

    @Test
    public void deleteBucketThrowsIfNotEmpty() {

        assertThrows(StorageException.class, () -> {
            try {
                client.deleteBucket("testbucket").get();
            } catch (ExecutionException e) {
                throw e.getCause();
            }
        }, "Storage not empty");
    }
}
