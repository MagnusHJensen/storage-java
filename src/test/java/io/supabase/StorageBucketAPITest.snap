io.supabase.StorageBucketAPITest.deleteBucket=[
  {
    "message": "Successfully deleted"
  }
]


io.supabase.StorageBucketAPITest.emptyBucket=[
  {
    "message": "Successfully emptied"
  }
]


io.supabase.StorageBucketAPITest.getBucket=[
  {
    "id": "bucket2",
    "name": "bucket2",
    "owner": "4d56e902-f0a0-4662-8448-a4d9e643c142",
    "isBucketPublic": false,
    "createdAt": "2021-02-17T04:43:32.770206+00:00",
    "updatedAt": "2021-02-17T04:43:32.770206+00:00",
    "bucketPublic": false
  }
]


io.supabase.StorageBucketAPITest.getBucketById=[
  {
    "id": "bucket2",
    "name": "bucket2",
    "owner": "4d56e902-f0a0-4662-8448-a4d9e643c142",
    "isBucketPublic": false,
    "createdAt": "2021-02-17T04:43:32.770206+00:00",
    "updatedAt": "2021-02-17T04:43:32.770206+00:00",
    "bucketPublic": false
  }
]


io.supabase.StorageBucketAPITest.updateBucket=[
  {
    "message": "Successfully updated"
  }
]